### PR TITLE
[4.0] Show changelog - GSoC Expand Extensions Manager

### PR DESCRIPTION
**Pull request by NunoLopesPT**
### Introduction

At this time, users have to look for external sources of information about changes in an extension, when they are going to update their extensions. In most cases, users install an extension without checking the changelog due to the difficulty to find the associated changelog. With this new feature, users can review the changelog of the extensions when they are about to install an update (granted that the changelog is provided by the extension's developer), and after the extension is installed.

### Summary of Changes

Added a new parameter on the xml updateserver files 'changelogurl' and adding it to #__updates table.

### Testing Instructions

I made some changes in the core databases so before testing you should fix your database:

screenshot from 2017-07-24 21-36-09

Extensions -> Database -> Fix
Extensions -> Updates -> Find Updates

 **Update View** 

Install my component: https://github.com/NunoLopes96/component_joomla 
By default you will have always an update from 0.0.2 to 4.0.0 just for testing

If the changelog is not available it will show N/A like in the Folder column

screenshot from 2017-05-23 18-28-40

It will open a modal with an iframe of the changelogurl in the body.
The title of the modal is version ' - ' component_name

screenshot from 2017-05-17 19-54-50

**Manage View**

The changelogurl is supported for Components, Libraries, Modules, Plugins and Packages, you just need to add <changelogurl> tags in your instalation xml with a valid url inside.

To test this you can use these modified packages:

(To test Package, Component, Module, Plugin)
[pkg_weblinks_3.4.1-changelog.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1127690/pkg_weblinks_3.4.1-changelog.zip)
screenshot from 2017-07-06 12-34-02

(To test Libraries)
[lib_fof30-3.1.0-changelog.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1127689/lib_fof30-3.1.0-changelog.zip)
screenshot from 2017-07-06 12-34-49

(To test languages)
The language installer its diferent from the others so you may want to test 2 cases:
1-Install without having it already in your extensions
2-Install without changelogurl and then install again with a changelogurl

This happens because it has 1 function to install and another one to update ( they are similar )

[admin_pt-PT.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1128171/admin_pt-PT.zip)
Extension name: Português (pt-PT)

(To test templates)
[et_kidtoy_joomla_template_install_package_free.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1128256/et_kidtoy_joomla_template_install_package_free.zip)
Extension name: etkidtoy

(To test Files)
[cli2zoombie.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1130860/cli2zoombie.zip)
Extension name: Zoombie - CLI Application

### Expected result

In Updates View you will have a new column that will show a changelog button to open a modal if there is a changelogurl in the #__updates table.
In Manage View you will have an <a> tag in the version with a info icon on the right side if your extension has a changelogurl in the #__extensions table.

### Documentation Changes Required

Add changelogurl parameter on updatesite information and in extension

### CODE REVIEW

**Comment by mbabker** [comment]
Looking at this with fresh eyes, there is a part of me that wonders if we should be using an iframe to show the third party site in this manner, or if we should implement a standardized file format (i.e. another XML schema) that developers publish which is consumed and rendered within that modal.  Beyond that, concept looks fine to me.

**Comment by brianteeman** [comment]
Using something like XML would be beneficial when extensions are stored on services that do not easily serve html

File: libraries/src/CMS/Installer/InstallerAdapter.php [hunk 1]
```diff
@@ -23,6 +23,14 @@
 abstract class InstallerAdapter
 {
 	/**
+	 * Changelog Url of an extensions
```
Comment (brianteeman) [inline libraries/src/CMS/Installer/InstallerAdapter.php]: Please URL in the text to satisfy the style guide and if it should be singular then remove the s. If it should be plural then delete the an

File: administrator/components/com_installer/tmpl/manage/default.php [hunk 2]
```diff
@@ -106,7 +106,29 @@
...
+												'<iframe src="' . $item->changelogurl . '"></iframe>');
+											?>
+										<?php else : ?>
+											<?php echo $item->version; ?>
+										<?php endif; ?>
+									<?php else:
+										echo '&#160;';
+									endif; ?>
+									<?php if ($item->changelogurl != null):?>
+										<i class="fa fa-info" aria-hidden="true"></i>
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/manage/default.php]: Please dont use the i tag for an icon - it is not correct markup. Instead please use the span tag

**Comment by conconnl** [comment]
The idea looks good but using a iFrame will create problems when using certain Security Headers.
Especially the header X-Frame-Options, this header will block loading a website in a Frame.
So when a extension developer use this header in his website, you won't see the information in a frame. https://scotthelme.co.uk/hardening-your-http-response-headers/#x-frame-options

So @mbabker his idea for having a standardized file format would be better.

**Comment by roland-d** [comment]
I have never been in favor of the iFrame, the issue mentioned by @conconnl underlines we better not use this. I will update this PR with a proposed XML scheme and we can shoot at it.

**Comment by roland-d** [comment]
So I have come up with the following schema. Let me know if I should proceed with this or if it should be changed.
```
<changelogs>
    <changelog>
        <element>com_lists</element>
        <type>component</type>
        <version>0.0.4</version>
        <security>
            <item></item>
            <item></item>
        </security>
        <fix>
            <item></item>
            <item></item>
        </fix>
        <language>
            <item></item>
            <item></item>
        </language>
        <addition>
            <item></item>
            <item></item>
        </addition>
        <change>
            <item></item>
            <item></item>
        </change>
        <removed>
            <item></item>
            <item></item>
        </removed>
        <note>
            <item></item>
            <item></item>
        </note>
    </changelog>
</changelogs>
```

This schema allows for the following types of information:
- Security fix
- Regular fix
- Language change
- Addition
- Change
- Remove
- Note

I believe this should cover all possible changes. Let me know what you think.

**Comment by ggppdk** [comment]
Looks good but we are talking about users (non-developers) reading this

maybe these categories

Security Fixes
Bug Fixes

New Features / Improvements
Removed features

Behaviour Change
Notes

**Comment by roland-d** [comment]
@ggppdk 
> but we are talking about users (non-developers) reading this

Yes, but these words are not the exact words I would use indeed and they will be language strings so they are open to what we are going to put in. We can take your text for the English language.

**Comment by roland-d** [comment]
I have started to bring this PR inline with the 4.0-dev codebase and work out the issues and improve the featureset. 

I am sure @Quy will help me with codestyle :) Now I also need help with styling the modal containing the updates. Have a look at this:
image

This contains all the possible sections that are available. @ciar4n is this something you can help me with?

As for the `changelog.xml` I am now using this format, any feedback for this is appreciated as well.
```
<?xml version="1.0" encoding="utf-8"?>
<changelogs>
    <changelog>
        <element>com_lists</element>
        <type>component</type>
        <version>0.0.4</version>
	<entries>
        <security>
            <item>CVE 1</item>
            <item>CVE 2</item>
        </security>
        <fix>
            <item>FIX 1</item>
            <item>FIX 2</item>
        </fix>
        <language>
            <item>LANG 1</item>
            <item>LANG 2</item>
        </language>
        <addition>
            <item>ADD 1</item>
            <item>ADD 2</item>
        </addition>
        <change>
            <item>CHANGE 1</item>
            <item>CHANGE 2</item>
        </change>
        <removed>
            <item>REMOVE 1</item>
            <item>REMOVE 2</item>
        </removed>
        <note>
            <item>NOTE 1</item>
            <item>NOTE 2</item>
        </note>
</entries>
    </changelog>
</changelogs>
```

This PR is not ready yet but I would appreciate feedback already so we can get this into a ready state.

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 3]
```diff
@@ -6,6 +6,14 @@
 COM_INSTALLER="Installer"
 COM_INSTALLER_AUTHOR_INFORMATION="Author Information"
 COM_INSTALLER_CACHETIMEOUT_LABEL="Updates Caching (in hours)"
+COM_INSTALLER_CHANGELOG="Changelog"
+COM_INSTALLER_CHANGELOG_SECURITY="Security Fixes"
+COM_INSTALLER_CHANGELOG_FIX="Bug Fixes"
+COM_INSTALLER_CHANGELOG_LANGUAGE="Changelog"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: why is this _LANGUAGE ?? or is it the "Changelog" that is wrong?

File: libraries/src/Installer/Adapter/ModuleAdapter.php [hunk 4]
```diff
@@ -257,7 +258,7 @@ protected function finaliseUninstall(): bool
 			}
 			catch (\RuntimeException $e)
 			{
-				\JLog::add(\JText::sprintf('JLIB_INSTALLER_ERROR_MOD_UNINSTALL_EXCEPTION', $e->getMessage()), \JLog::WARNING, 'jerror');
+				\JLog::add(Text::sprintf('JLIB_INSTALLER_ERROR_MOD_UNINSTALL_EXCEPTION', $e->getMessage()), \JLog::WARNING, 'jerror');
```
Comment (brianteeman) [inline libraries/src/Installer/Adapter/ModuleAdapter.php]: please namespacce JLog

+use Joomla\CMS\Log\Log;

File: administrator/language/en-GB/en-GB.com_installer.ini (hunk 3 above)
Comment (roland-d) [reply administrator/language/en-GB/en-GB.com_installer.ini]: It is actually "Changelog" that is wrong. This needs to be "Language" as it is about language file changes listed in the changelog.

Comment (brianteeman) [reply administrator/language/en-GB/en-GB.com_installer.ini]: Thanks

File: administrator/components/com_installer/tmpl/update/default.php [hunk 5]
```diff
@@ -25,106 +28,130 @@
...
+								</tr>
 							</thead>
 							<tfoot>
-							<tr>
-								<td colspan="9">
-									<?php echo $this->pagination->getListFooter(); ?>
-								</td>
-							</tr>
+								<tr>
+									<td colspan="9">
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/update/default.php]: Change to `10`

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 6]
```diff
@@ -6,6 +6,14 @@
 COM_INSTALLER="Installer"
 COM_INSTALLER_AUTHOR_INFORMATION="Author Information"
 COM_INSTALLER_CACHETIMEOUT_LABEL="Updates Caching (in hours)"
+COM_INSTALLER_CHANGELOG="Changelog"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_installer.ini:9]: Sort keys in alpha order

File: libraries/src/Installer/Adapter/LanguageAdapter.php [hunk 7]
```diff
@@ -878,21 +882,21 @@ public function discover_install()
...
+		$manifest_details                        = Installer::parseXMLInstallFile($this->parent->getPath('manifest'));
 		$this->parent->extension->manifest_cache = json_encode($manifest_details);
-		$this->parent->extension->name = $manifest_details['name'];
+		$this->parent->extension->name           = $manifest_details['name'];
 
 		if ($this->parent->extension->store())
 		{
 			return true;
 		}
 		else
```
Comment (Quy) [inline libraries/src/Installer/Adapter/LanguageAdapter.php]: No need to be in `else` statement.

File: administrator/components/com_installer/tmpl/manage/default.php [hunk 8]
```diff
@@ -78,61 +80,102 @@
...
+										</label>
+									</td>
+									<td class="text-center">
+										<?php echo $item->client_translated; ?>
+									</td>
+									<td class="text-center">
+										<?php echo $item->type_translated; ?>
+									</td>
+									<td class="d-none d-md-table-cell text-center">
+										<?php if ($item->version != ''): ?>
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/manage/default.php]: Add space before `:`

File: administrator/components/com_installer/tmpl/manage/default.php [hunk 9]
```diff
@@ -78,61 +80,102 @@
...
+											<?php if ($item->changelogurl != null): ?>
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/manage/default.php]: Add space before `:`

File: administrator/components/com_installer/tmpl/update/default.php [hunk 10]
```diff
@@ -25,106 +28,130 @@
...
+											<?php echo $item->type_translated; ?>
+										</td>
+										<td class="d-none d-md-table-cell text-center">
+											<span class="badge badge-warning"><?php echo $item->current_version; ?></span>
+										</td>
+										<td class="d-none d-md-table-cell text-center">
+											<span class="badge badge-success"><?php echo $item->version; ?></span>
+										</td>
+										<td class="d-none d-md-table-cell text-center">
+											<?php if ($item->changelogurl != null):?>
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/update/default.php]: Add space before and after `:`

File: administrator/components/com_installer/tmpl/update/default.php [hunk 11]
```diff
@@ -25,106 +28,130 @@
...
+												'changelog_modal' . $item->extension_id,
+												array(
+													'title' => $item->version . " - " . $item->name,
+													'bodyHeight'  => '60',
+													'modalWidth'  => '60',
+												),
+												'<iframe src="' . $item->changelogurl . '"></iframe>');
+											?>
+
+											<?php else:?>
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/update/default.php]: Add space before and after `:`

File: administrator/components/com_installer/tmpl/update/default.php [hunk 12]
```diff
@@ -25,106 +28,130 @@
...
+											<span>
+												<?php echo Text::_('COM_INSTALLER_TYPE_NONAPPLICABLE')?>
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/update/default.php]: Add `;` and space

File: administrator/language/en-GB/en-GB.com_installer.ini (hunk 6 above)
Comment (roland-d) [reply administrator/language/en-GB/en-GB.com_installer.ini:9]: @Quy What is not sorted here? CA is before CH.

Comment (Quy) [reply administrator/language/en-GB/en-GB.com_installer.ini:9]: Sort keys from lines 10 - 16.

Comment (roland-d) [reply administrator/language/en-GB/en-GB.com_installer.ini:9]: All done.

**Comment by wilsonge** [comment]
I don't think you need the `type` in your schema. This living on a foreign server where you are already matching against the `element` so I don't see what value it's providing?

Additionally I think we only need `fix` and `note` - not `change` - they should be covered by the other things

**Comment by roland-d** [comment]
@wilsonge The reason the `type` is there is because we have it that way in the update XMLs as well. The only reason I took it as well is because I figured we can use it for better checking we have the correct type.

> Additionally I think we only need fix and note - not change - they should be covered by the other things

As an extension developer I use all of them to clearly identify where the entry belongs to. Each has their own merit to me at least. I don't see any harm in offering them, extension developers don't have to use them of course. They are there for those who want to use more distinction.

**Comment by brianteeman** [comment]
Agree with roland - hardly  changelog without listing any changes ;)

**Comment by roland-d** [comment]
Closing this as I can't push to the repo anymore. Continuing in #24026

**Pull request closed without merging**